### ISSUE TITLE
Inappropriate transistors

### ISSUE THREAD

**Issue by Harvie** [carneeki/OpenVFD#2]
Hi! according to Block Diagram.png you are using IRF512 transistors, but i've checked datasheet and these came in 100V maximum rating. I guess that commercial VFDs use IGBT transistors... Don't know.

But you surely get more  than 100V when rectifiing mains AC, no matter if US 120V or EU 230V...  
Also 230V is RMS, while after rectification you get peak value 320V rather than the RMS. (multiply by factor of 1.41421)
But if you are rectifiing 3 phases, that's different story. There is 400V RMS between two phases, which means peak is something like 566V, so that's what you get after rectifiing 3F mains in EU. (multiply by 1.5708)

So to know what voltage you THEORETICALY get after rectification of 3 phases:

**EU: 230V * 1.5708 * 1.41421 = 511V
US: 120V * 1.5708 * 1.41421 = 267V**

Note that the voltage in grid fluctuates a little. Also you should leave some safety margin.

Please keep that in mind.

Anyway... Cool project! I wonder how much it takes to get this working and if it will be cheaper to make localy than buying chinese VFD :-)

**Comment by carneeki** [comment]
@Harvie absolutely correct on the count of voltage, nice find. :)

IIRC, the block diagram was the very first diagram I did to get a "big picture" overview - the idea being that each area of the diagram would have separate schematics with values, and I likely forgot to remove the component values from the block diagram.

I seem to remember another IRF branded IGBT in the output schematic with a voltage rating of 600V, and thinking that 90V for AU (also 230V, +8%/-10%) wasn't a lot of margin [in theory could be about 680V for our power here]... However, there are two IGBTs in series (as well as the motor) along each path from +ve to GND along the DC bus, so it works out to about 300V across the drain and source on each IGBT.

Out of curiosity, where does 1.5708 come from? I recognise 1.414... = sqrt(2), and 1.732... = sqrt(3) which were both common constants in 3ph power calculations, but 1.5708... is a new one to me.

Thanks again for spotting this :)

**Comment by Harvie** [comment]
> Out of curiosity, where does 1.5708 come from?

I am not really sure, since i've never designed 3 phase appliance. I've just googled the number and 1.57 seems to repeat through te internet. However i draw these two green arrows to the following diagram. I seems to be more like 1.5 or 1.49 in picture, i don't know why everybody uses 1.57. Maybe i am missing something. But i guess there's more to it:

![image](https://user-images.githubusercontent.com/276504/54343710-b2e8dc80-463f-11e9-93f7-f728fa206509.png)

(left arrow is 1P to ground, second arrow is between two phases)

> However, there are two IGBTs in series (as well as the motor) along each path from +ve to GND along the DC bus

I don't think so:

![image](https://user-images.githubusercontent.com/276504/54344250-b3ce3e00-4640-11e9-8e66-eb0cd2adab98.png)

If you want possitive voltage on U1, you have to open Q1 and close Q4. That would mean that Q4 gets full DC rail voltage.

Also you can't assume anything about the motor. VFDs should be able to work with low-impedance low-voltage motors for several reasons.

1.) when you lower the RPM, you lower the frequency and the impedance goes very low. That's why common VFDs for CNC spindles have to dramaticaly lower the sine voltage at the low RPMs. That's why all VFDs have programmable voltage/frequency curves.

2.) you might want VFD to drive small motor that can't handle 230V at all. For example you might want to drive small BLDC motor using VFD.

3.) VFD should have closed-loop current sensing. I've accidentaly managed to get some metalic chips to the connector of my CNC spindle and that effectively created short circuit of VFD. I should have been more carefull, but to be honest the metalic shavings are very common around CNC.
I am very happy that VFD managed to detect the short circuit, stop itself and display appropriate error message instead of exploding transistors...

**Comment by carneeki** [comment]
> But i guess there's more to it:
Perhaps this image will help?
![image](https://user-images.githubusercontent.com/481448/54349603-bd3ac300-469f-11e9-8eda-b869ea9a2b6d.png)

It's from the section of [Full-wave Three-phase Rectifier Conduction Waveform](https://www.electronics-tutorials.ws/power/three-phase-rectification.html) here which offers an explanation of a slightly higher than 1.57x constant (approx 1.65) is needed. Overlayed with a variation on your image I think it makes sense:
![image](https://user-images.githubusercontent.com/481448/54350000-831df100-46a0-11e9-96e5-a677527d4eef.png)

I took your P2 to P1 voltage and copy+pasted it next to the yellow line I drew and the yellow line is longer just a little bit either side of the valley - the gradient of P3 is near zero so the rate of change is also near zero, but the gradient of P2 is closer to 1 so the rate of change is more. (btw, I honestly never thought I'd use calculus in a discussion on the internet :D !)

> If you want possitive voltage on U1, you have to open Q1 and close Q4. That would mean that Q4 gets full DC rail voltage.
Yep - this is true... It makes more sense now that I see the H-bridge on paper again, and to pretend the IGBTs are switches.

Concerning:
1. Lower voltage - I think they simply use PWM to do this in software no? Would be interesting to put an oscilloscope on the output terminals of the FETs to take a look. If so, this become a matter of software chopping the open/close ratios.
2. Low voltage DC would also be a matter of software chopping PWM, but maybe with additional timing constraints - I've never looked at BLDC control before.
3. The output schematic includes an IRS2336, the [datasheet](https://www.infineon.com/dgdl/Infineon-IRS2336-DS-v02_00-EN.pdf?fileId=5546d462533600a40153567ab2a0280e) indicates it has short and shoot-thru protections, as well as fault reporting. It requires hooking up the Itrip pin to some shunt resistors. I'll add that to the growing TODO list :)

**Comment by Harvie** [comment]
> I think they simply use PWM to do this in software no? 

Of course. But that was not my point. My point is that in such case you have to drop all DC rail voltage on the transistor, so you can't count with motor to drop any aditional voltage to save your transistors.

> Low voltage DC would also be a matter of software chopping PWM.  I've never looked at BLDC control before.

BLDC is exactly the same as 3 phase AC motor. The **DC** in BLDC stands just for the fact that they use "speed controller" to run on DC. Speed controller is basicaly just VFD without rectifier on input. BLDC terminology is used by RC model people (drones, rc airplanes, e-bikes, etc...), so that means BLDC motor usualy runs much lower voltages, because they use batteries to power them. Technicaly speaking it's exactly the same as AC motor. The BLDC term is bit misleading... But if you think about it... Even brushed DC motors run on AC. They just have to create it using brushes.

Here is how it looks. Both motor and speed controller:

![image](https://user-images.githubusercontent.com/276504/54352665-75417f00-4652-11e9-929a-7f1bba67b52b.png)

But note that "kv" doesn't mean "kilovolts". It means "RPMs per volt". So this one can get 1000RPMs when running without load at 1V (or 2000RPM at 2V, etc...).

I've never tried to run such motor using my 230V huanyang VFD, but i think it should be possible, because i can dial the voltage down to few volts.

BTW You can get such motors for experimenting from old harddrives...

> The output schematic includes an IRS2336, the datasheet indicates it has short and shoot-thru protections, as well as fault reporting.

This is very cool! At least if that's really true... Is it really protecting short through the IGBTs? Or it's just protecting the driver output from being shorted?
